File: backend/app/schemas/ingestion.py
import uuid
from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, ConfigDict, Field


class DemoIngestRequest(BaseModel):
    dataset_scale: str = Field(default="enterprise_250k", description="Scale of demo dataset: express_10k | standard_50k | enterprise_250k")
    include_fraud_scenarios: bool = Field(default=True, description="Include realistic attack vectors: Card testing, Account takeover, Velocity spike, etc.")
    auto_run_ai_pipeline: bool = Field(default=True, description="Automatically calculate features, AI models, decisions, and graph links")


class DemoIngestStageResult(BaseModel):
    stage: str
    status: str = Field(default="COMPLETED")
    count: int
    duration_ms: float
    description: str


class DemoIngestResponse(BaseModel):
    batch_id: str
    scale: str
    merchants_count: int
    customers_count: int
    devices_count: int
    transactions_count: int
    features_count: int
    predictions_count: int
    decisions_count: int
    cases_count: int
    graph_nodes_count: int
    graph_edges_count: int
    notifications_count: int
    stages: List[DemoIngestStageResult]
    total_duration_ms: float
    summary_message: str


class FileParseResponse(BaseModel):
    filename: str
    file_type: str
    extracted_files: List[str] = Field(default_factory=list)
    detected_entity_type: str
    total_rows: int
    total_columns: int
    columns: List[str]
    detected_schema: Dict[str, str]
    quality_score: float
    missing_values_count: int
    duplicate_rows_count: int
    validation_warnings: List[str] = Field(default_factory=list)
    sample_records: List[Dict[str, Any]] = Field(default_factory=list)


class BatchIngestRequest(BaseModel):
    entity_type: str = Field(default="TRANSACTIONS", description="TRANSACTIONS | CUSTOMERS | MERCHANTS | DEVICES")
    filename: str
    records: List[Dict[str, Any]]
    column_mapping: Optional[Dict[str, str]] = None
    run_ai_pipeline: bool = Field(default=True)


class BatchIngestResponse(BaseModel):
    import_id: str
    entity_type: str
    total_received: int
    successfully_imported: int
    skipped_records: int
    features_generated: int
    ai_predictions_evaluated: int
    decisions_produced: int
    cases_opened: int
    duration_ms: float
    status: str
    message: str


class ImportHistoryItem(BaseModel):
    import_id: str
    filename: str
    source_type: str
    entity_type: str
    rows_processed: int
    rows_skipped: int
    quality_score: float
    duration_ms: float
    status: str
    created_at: datetime
    imported_by: str


class PaginatedImportHistoryResponse(BaseModel):
    items: List[ImportHistoryItem]
    total: int
    page: int
    size: int

File: frontend/src/types/ingestion.ts
export interface DemoIngestRequest {
  dataset_scale: "enterprise_250k" | "standard_50k" | "express_10k";
  include_fraud_scenarios: boolean;
  auto_run_ai_pipeline: boolean;
}

export interface DemoIngestStageResult {
  stage: string;
  status: "COMPLETED" | "IN_PROGRESS" | "FAILED";
  count: number;
  duration_ms: number;
  description: string;
}

export interface DemoIngestResponse {
  batch_id: string;
  scale: string;
  merchants_count: number;
  customers_count: number;
  devices_count: number;
  transactions_count: number;
  features_count: number;
  predictions_count: number;
  decisions_count: number;
  cases_count: number;
  graph_nodes_count: number;
  graph_edges_count: number;
  notifications_count: number;
  stages: DemoIngestStageResult[];
  total_duration_ms: number;
  summary_message: string;
}

export interface FileParseResponse {
  filename: string;
  file_type: string;
  extracted_files: string[];
  detected_entity_type: string;
  total_rows: number;
  total_columns: number;
  columns: string[];
  detected_schema: Record<string, string>;
  quality_score: number;
  missing_values_count: number;
  duplicate_rows_count: number;
  validation_warnings: string[];
  sample_records: Record<string, any>[];
}

export interface BatchIngestRequest {
  entity_type: string;
  filename: string;
  records: Record<string, any>[];
  column_mapping?: Record<string, string>;
  run_ai_pipeline: boolean;
}

export interface BatchIngestResponse {
  import_id: string;
  entity_type: string;
  total_received: number;
  successfully_imported: number;
  skipped_records: number;
  features_generated: number;
  ai_predictions_evaluated: number;
  decisions_produced: number;
  cases_opened: number;
  duration_ms: number;
  status: string;
  message: string;
}

export interface ImportHistoryItem {
  import_id: string;
  filename: string;
  source_type: string;
  entity_type: string;
  rows_processed: number;
  rows_skipped: number;
  quality_score: number;
  duration_ms: number;
  status: string;
  created_at: string;
  imported_by: string;
}

export interface PaginatedImportHistoryResponse {
  items: ImportHistoryItem[];
  total: number;
  page: number;
  size: number;
}
